bgpd/bgp_mplsvpn.c

```diff
@@ -233,6 +236,762 @@ int bgp_nlri_parse_vpn(struct peer *peer, struct attr *attr,
+	struct bgp_node *bn;
+	const char *debugmsg;
+
+	if (debug)
+		zlog_debug("%s: debug=%d, info_vrf->type=%d", __func__, debug,
+			   info_vrf->type);
+
+	if (info_vrf->type == ZEBRA_ROUTE_BGP_VPN) {
+		/* loop check! */
+		return;
+	}
+	if (!bgp_vpn)
+		return;
+
+	if (!afi) {
+		if (debug)
+			zlog_debug("%s: can't get afi of prefix", __func__);
+		return;
+	}
+
+	if (!vpn_leak_to_vpn_active(bgp_vrf, afi, &debugmsg)) {
+		if (debug)
+			zlog_debug("%s: skipping: %s", __func__, debugmsg);
+		return;
+	}
+
+	bgp_attr_dup(&static_attr, info_vrf->attr); /* shallow copy */
+
+	/*
+	 * route map handling
+	 */
+	if (bgp_vrf->vpn_policy[afi].rmap[BGP_VPN_POLICY_DIR_TOVPN]) {
+		struct bgp_info info;
+		route_map_result_t ret;
+
+		memset(&info, 0, sizeof(info));
+		info.peer = info_vrf->peer;
+		info.attr = &static_attr;
+		ret = route_map_apply(
+			bgp_vrf->vpn_policy[afi].rmap[BGP_VPN_POLICY_DIR_TOVPN],
+			p, RMAP_BGP, &info);
+		if (RMAP_DENYMATCH == ret) {
+			bgp_attr_flush(&static_attr); /* free any added parts */
+			if (debug)
+				zlog_debug(
+					"%s: vrf %s route map \"%s\" says DENY, returning",
+					__func__, bgp_vrf->name,
+					bgp_vrf->vpn_policy[afi]
+						.rmap[BGP_VPN_POLICY_DIR_TOVPN]
+						->name);
+			return;
+		}
+	}
+
+	/* TBD should we preserve or ignore ecommunity values from info_vrf? */
+	if (static_attr.ecommunity) {
+		static_attr.ecommunity = ecommunity_merge(
+			static_attr.ecommunity,
+			bgp_vrf->vpn_policy[afi]
+				.rtlist[BGP_VPN_POLICY_DIR_TOVPN]);
+	} else {
+		static_attr.ecommunity = ecommunity_dup(
+			bgp_vrf->vpn_policy[afi]
+				.rtlist[BGP_VPN_POLICY_DIR_TOVPN]);
+	}
+	if (static_attr.ecommunity->size) {
+		SET_FLAG(static_attr.flag,
+			 ATTR_FLAG_BIT(BGP_ATTR_EXT_COMMUNITIES));
+	} else {
+		ecommunity_free(&static_attr.ecommunity);
+	}
+
+	/* Nexthop */
+	/* if nexthop not set, use router-id */
+	if (CHECK_FLAG(bgp_vrf->vpn_policy[afi].flags,
+		       BGP_VPN_POLICY_TOVPN_NEXTHOP_SET)) {
+
+		struct prefix *nexthop =
+			&bgp_vrf->vpn_policy[afi].tovpn_nexthop;
+		switch (nexthop->family) {
+		case AF_INET:
+			/* prevent mp_nexthop_global_in <- self in bgp_route.c
+			 */
+			static_attr.nexthop.s_addr = nexthop->u.prefix4.s_addr;
+
+			static_attr.mp_nexthop_global_in = nexthop->u.prefix4;
+			static_attr.mp_nexthop_len = 4;
+			break;
+
+		case AF_INET6:
+			static_attr.mp_nexthop_global = nexthop->u.prefix6;
+			static_attr.mp_nexthop_len = 16;
+			break;
+
+		default:
+			assert(0);
+		}
+	} else {
+		static_attr.nexthop = bgp_vrf->router_id;
+		static_attr.mp_nexthop_global_in = bgp_vrf->router_id;
+		static_attr.mp_nexthop_len = 4;
+	}
+
+	if (CHECK_FLAG(bgp_vrf->vpn_policy[afi].flags,
+		       BGP_VPN_POLICY_TOVPN_LABEL_SET)) {
+
+		label_val = bgp_vrf->vpn_policy[afi].tovpn_label;
+	} else {
+		/* TBD get from label manager */
+		label_val = MPLS_LABEL_IMPLICIT_NULL;
+	}
+
+	/* Set originator ID to "me" */
+	SET_FLAG(static_attr.flag, ATTR_FLAG_BIT(BGP_ATTR_ORIGINATOR_ID));
```

Comment:
I thought the use/need of the Originator ID is only for Route Reflection. Why is it being set here?

Reply:
I believe that the originator id is only included in the outgoing update when bgpd is acting as a RR:

  ```
   bgp_attr.c:
   3078         /* Route Reflector. */
   3079         if (peer->sort == BGP_PEER_IBGP && from
   3080             && from->sort == BGP_PEER_IBGP) {
   3081                 /* Originator ID. */
   3082                 stream_putc(s, BGP_ATTR_FLAG_OPTIONAL);
   3083                 stream_putc(s, BGP_ATTR_ORIGINATOR_ID);
   3084                 stream_putc(s, 4);
   3085
   3086                 if (attr->flag & ATTR_FLAG_BIT(BGP_ATTR_ORIGINATOR_ID))
   3087                         stream_put_in_addr(s, &attr->originator_id);
   3088                 else
   3089                         stream_put_in_addr(s, &from->remote_id);
```

imported/exported routes still have their original peer value, which can lead to the wrong originator id when we act as RR (problem we encountered some years ago in the vnc/rfapi code). Always setting it seems to do the right thing. Does this approach break anything?

Reply:
It seems okay to always set the originator id -- it is generally ignored except in the case of an RR. Might be worth having @dwalton76 comment here, however, to be certain.